### PR TITLE
Move half traits to private header and add half/bhalf infinity trait

### PR DESCRIPTION
**Pull request by e10harvey**
## Summary of changes
- core/src: Move half traits to private header
- core/src: Add half_t and bhalf_t infinity trait

Related to #6045 and https://github.com/kokkos/kokkos-kernels/pull/1774.

### CODE REVIEW

File: core/src/Kokkos_Half_NumericTraits.hpp
Comment (dalg24) [inline core/src/Kokkos_Half_NumericTraits.hpp:1]: If as the title line of your PR suggest this header is intended to be a private header, it is probably better to move it to the impl/ subdirectory.

**Review by dalg24 — COMMENTED** [review]
Please update the title line of your PR and make sure it mentions this is adding the half precision specializations of the infinity traits that were missing.
This is the most relevant of your PR because that is the only part that is meant to be observable by the user.

**Review by dalg24 — APPROVED** [review]

File: core/src/impl/Kokkos_Half_NumericTraits.hpp [hunk 1]
```diff
@@ -0,0 +1,349 @@
...
+#if defined(KOKKOS_HALF_T_IS_FLOAT) && !KOKKOS_HALF_T_IS_FLOAT
+/// \brief: Infinity
+///
+/// Binary16 encoding:
+///             [s  e  e  e  e  e  f f f f f f f f f f]
+///             [0  1  1  1  1  1  0 0 0 0 0 0 0 0 0 0]
+/// bit index:   15 14 13 12 11 10 9 8 7 6 5 4 3 2 1 0
+///
+template <>
+struct Kokkos::Experimental::Impl::infinity_helper<Kokkos::Experimental::half_t> {
```
Comment (dalg24) [inline core/src/impl/Kokkos_Half_NumericTraits.hpp:72]: you are relying on transitive includes for `half_t`

**Review by nliber — APPROVED** [review]

**Review by masterleinad — CHANGES_REQUESTED** [review]
Please split the one commit in such a way that one commit solely moves code without changing anything and other commits fix typos and add stuff. Otherwise, it is very hard to see what the real changes in this pull request are.

**Comment by dalg24** [comment]
Just posting it here in case it saves some time to other reviewers
The 1st commit is not per se a move (I pasted the diff below) but I suppose it is ok.  `develop` has comments about the infinity trait and the actual implementation had been lost somewhere in translation.
```diff
--- develop
+++ this_pr
@@ -39,17 +39,7 @@
 //           2**15 * (1 + 0.9990234375) =
 //           65504.0
 //
-
-/// \brief: Infinity.
-///
-/// base2 encoding: bits [10,14] set
-/// #define KOKKOS_IMPL_HALF_T_HUGE_VALH 0x7c00
-/// Binary16 encoding:
-///             [s  e  e  e  e  e  f f f f f f f f f f]
-///             [0  1  1  1  1  1  0 0 0 0 0 0 0 0 0 0]
-/// bit index:  15 14 13 12 11 10 9 8 7 6 5 4 3 2 1 0
 #if defined(KOKKOS_HALF_T_IS_FLOAT) && !KOKKOS_HALF_T_IS_FLOAT
-
 /// \brief: Minimum normalized number
 ///
 /// Stdc defines this as the smallest number (representable in binary16).
@@ -194,7 +184,7 @@

 /// \brief: This is the smallest possible exponent value
 ///
-/// Stdc defines this as the smallest possible exponent value for type binary16.
+/// Stdc defines this as the smallest possible exponent val`ue for type binary16.
 /// More precisely, it is the minimum negative integer such that the value min_exponent_helper
 /// raised to this power minus 1 can be represented as a normalized floating point number of type float.
 ///
@@ -309,4 +299,4 @@
   static constexpr int value = 128;
 };
 #endif
-////////////// END BHALF_T (bfloat16) limits //////////////
+////////////// END BHALF_T (bfloat16) limits //////////
```

Please expand the description.  Say something like:
```
This PR does 2 things
* move the numeric traits specializations for `half_t` and `bhalf_t` to a new `<core/src/impl/Kokkos_Half_NumericTraits.hpp>` header
* add the infinity trait for the half types that was missing (oversight)
```

File: core/src/impl/Kokkos_Half_NumericTraits.hpp [hunk 2]
```diff
@@ -0,0 +1,349 @@
...
+///
+/// Stdc defined this as the value of the base, or radix, of the exponent representation.
+template <>
+struct Kokkos::Experimental::Impl::radix_helper<Kokkos::Experimental::half_t> {
+  static constexpr int value = 2;
+};
+
+/// \brief: This is the smallest possible exponent value
+///
+/// Stdc defines this as the smallest possible exponent val`ue for type binary16.
```
Comment (dalg24) [inline core/src/impl/Kokkos_Half_NumericTraits.hpp:220]: Fix the typo that was introduced in the splitting header up
```suggestion
/// Stdc defines this as the smallest possible exponent value for type binary16. 
```

**Review by masterleinad — APPROVED** [review]
Looks OK to me now (pending Damien's comments).

**Comment by dalg24** [comment]
Failures are unrelated (HIP build timing out)
Counting Daniel's approval which is not showing since a new review was requested.

**Review by dalg24 — APPROVED** [review]

**Pull request merged**